project-koku/koku
# [COST-4110] Handle duplicate usage ids from subs processing

## Jira Ticket

[COST-4110](https://issues.redhat.com/browse/COST-4110)

## Description

This change will adjust SUBS processing to only send information for one AWS usage account per customer. If a source is deleted and a different source has the same usage account id, it will take over sending data for that usage account.

## Testing

1. Checkout Branch
2. Restart Koku with `make docker-up-min-with-subs` to create a subs worker
3. Create a few nise reports with the same usage account id. sample yml (remove the .txt as github doesn't allow ymls): [subs-nise-demo.yml.txt](https://github.com/project-koku/koku/files/12643973/subs-nise-demo.yml.txt)
```
nise report aws  --aws-s3-bucket-name testing/local_providers/aws_local_1 --aws-s3-report-name report --static-report-file subs-nise-demo.yml
nise report aws  --aws-s3-bucket-name testing/local_providers/aws_local_2 --aws-s3-report-name report --static-report-file subs-nise-demo.yml
nise report aws  --aws-s3-bucket-name testing/local_providers/aws_local_3 --aws-s3-report-name report --static-report-file subs-nise-demo.yml
```
4. Create a source for the first bucket and ingest the data.
5. Watch the logs in the subs worker and see that line items were identified for metered RHEL and check minio for a subs report to have been generated in a path similar to this:
```
koku-bucket/org1234567/AWS/source=091b3d05-a6fd-4848-8fd5-449cf61b9dd5/date=2023-09-17/subs_cb5887a3-d1e1-43f7-a7c6-4a983709438d_0.csv
```
6. Create a source for the second bucket and ingest the data
7. Watching the logs this time, you should see the following message: `no valid IDs to process for current source.`
8. Check that no new report was uploading to the subs path in the bucket in minio
9. Delete the first source
10. Create a source for the third bucket and ingest the data
11. The logs should indicate line items were identified for metered RHEL and a new report should be in the bucket at a path similar to above.
12. Check postgres for the subs_id_map table and verify the correct usage id is matched with the latest source data was processed for:
```
postgres=# select * from org1234567.reporting_subs_id_map;
 id |   usage_id    |             source_uuid              
----+---------------+--------------------------------------
  7 | 9999999999998 | 091b3d05-a6fd-4848-8fd5-449cf61b9dd5
```


## Notes

...
